libsolidity/analysis/TypeChecker.cpp

```diff
@@ -335,6 +332,12 @@ void TypeChecker::endVisit(ModifierDefinition const& _modifier)
 
 bool TypeChecker::visit(FunctionDefinition const& _function)
 {
+	if (_function.isFree())
+	{
+		m_freeFunctionDefinitions[_function.name()].push_back(&_function);
+		ContractLevelChecker{m_errorReporter}.findDuplicateDefinitions<FunctionDefinition const*>(m_freeFunctionDefinitions);
```

Comment:
I meant this should be a feature of the ContractLevelChecker in the following way:
Change the `check` function of ContractLevelChecker such that it gets a SourceUnit instead of a Contract and move the inner loop from CompilerStack into ContractLevelChecker.

The main problem I see there is that we maybe don't have the information anymore that as generated in the ReferencesResolver about which symbols are imported into which scopes.